viz/cmd/profile.go

```diff
@@ -0,0 +1,213 @@
+package cmd
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/ghodss/yaml"
+	"github.com/linkerd/linkerd2/controller/api/util"
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha2"
+	pkgcmd "github.com/linkerd/linkerd2/pkg/cmd"
+	"github.com/linkerd/linkerd2/pkg/healthcheck"
+	"github.com/linkerd/linkerd2/pkg/k8s"
+	"github.com/linkerd/linkerd2/pkg/profiles"
+	"github.com/linkerd/linkerd2/pkg/protohttp"
+	"github.com/linkerd/linkerd2/pkg/tap"
+	pb "github.com/linkerd/linkerd2/viz/metrics-api/gen/viz"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation"
+)
+
+type profileOptions struct {
+	name          string
+	namespace     string
+	tap           string
+	tapDuration   time.Duration
+	tapRouteLimit uint
+}
+
+func newProfileOptions() *profileOptions {
+	return &profileOptions{
+		tapDuration:   5 * time.Second,
+		tapRouteLimit: 20,
+	}
+}
+
+func (options *profileOptions) validate() error {
+	if options.tap == "" {
+		return errors.New("The --tap flag must be specified")
+	}
+	// a DNS-1035 label must consist of lower case alphanumeric characters or '-',
+	// start with an alphabetic character, and end with an alphanumeric character
+	if errs := validation.IsDNS1035Label(options.name); len(errs) != 0 {
+		return fmt.Errorf("invalid service %q: %v", options.name, errs)
+	}
+	// a DNS-1123 label must consist of lower case alphanumeric characters or '-',
+	// and must start and end with an alphanumeric character
+	if errs := validation.IsDNS1123Label(options.namespace); len(errs) != 0 {
+		return fmt.Errorf("invalid namespace %q: %v", options.namespace, errs)
+	}
+	return nil
+}
+
+// newCmdProfile creates a new cobra command for the Profile subcommand which
+// generates Linkerd service profile based off tap data.
+func newCmdProfile() *cobra.Command {
+	options := newProfileOptions()
+
+	cmd := &cobra.Command{
+		Use:   "profile [flags] --tap resource (SERVICE)",
+		Short: "Output service profile config for Kubernetes based off tap data",
+		Long:  "Output service profile config for Kubernetes based off tap data.",
+		Example: `  # Generate a profile by watching live traffic.
+  linkerd profile -n emojivoto web-svc --tap deploy/web --tap-duration 10s --tap-route-limit 5
+`,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if options.namespace == "" {
+				options.namespace = pkgcmd.GetDefaultNamespace(kubeconfigPath, kubeContext)
+			}
+			options.name = args[0]
+			clusterDomain := "cluster.local"
```

Comment:
Better to have this is a const in `viz/cmd/root.go` so that this can be re-used? just like the core cmd.

Reply:
Yep this is added. Replace a use of it in `metrics-api/cmd` as well.

Reply:
I backed these commits out. There are docker build errors when trying to pipe this through to metrics-api and then we end up only using it in `profile.go`. Because it's only used there, it doesn't seem worth moving it to `root.go` since it's not duplicated anywhere else.